File: test_project/gen_code_cs/tbshape.cs
namespace demo
{

    public class TbShape
    {
        private readonly System.Collections.Generic.Dictionary<int, Shape> _dataMap;
        private readonly System.Collections.Generic.List<Shape> _dataList;
        
        public TbShape(System.Collections.Generic.List<Shape> dataList)
        {
            _dataList = dataList;
            _dataMap = new System.Collections.Generic.Dictionary<int, Shape>();
            foreach(var v in _dataList)
            {
                _dataMap.Add(v.Id, v);
            }
        }
        
        public System.Collections.Generic.Dictionary<int, Shape> DataMap => _dataMap;
        public System.Collections.Generic.List<Shape> DataList => _dataList;
        
        public Shape GetOrDefault(int key)
        {
            if (_dataMap.TryGetValue(key, out var v))
            {
                return v;
            }
            return null;
        }
        
        public Shape Get(int key)
        {
            if (_dataMap.TryGetValue(key, out var v))
            {
                return v;
            }
            throw new System.Collections.Generic.KeyNotFoundException();
        }
    }

}

File: test_project/gen_code_ts/tbshape.ts
export namespace demo {

    export class TbShape {
    private readonly _dataMap: Map<number, Shape>;
    private readonly _dataList: Shape[];
    
    constructor(dataList: Shape[]) {
        this._dataList = dataList;
        this._dataMap = new Map<number, Shape>();
        for (const v of this._dataList) {
            this._dataMap.set(v.Id, v);
        }
    }
    
    public get dataMap(): Map<number, Shape> {
        return this._dataMap;
    }
    
    public get dataList(): Shape[] {
        return this._dataList;
    }
    
    public get(key: number): Shape | undefined {
        return this._dataMap.get(key);
    }
}

}
